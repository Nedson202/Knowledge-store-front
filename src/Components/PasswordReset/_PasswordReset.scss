.password-reset {
  max-width: 400px;
  width: 100%;
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);

  .form-group {
    margin: auto;
    margin-bottom: 20px !important;

    li {
      list-style-type: none;

      i {
        position: absolute;
        top: 33px;
        right: 10px;
      }
    }
  }

  &__button {
    display: flex;
    justify-content: center;
    margin-top: 25px !important;
  }
}