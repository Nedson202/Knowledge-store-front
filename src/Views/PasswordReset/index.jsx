import queryString from 'querystring';
import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { compose, graphql } from 'react-apollo';
import { connect } from 'react-redux';
import setQuery from 'set-query-string';

import { resetPassword } from '../../queries/auth';
import { passwordToggler, tokenDecoder, toaster } from '../../utils';
import { setCurrentUser } from '../../redux/actions/userActions';
import {
  RESET_PASSWORD_QUERY, TOASTR_ERROR, MY_BOOKS_PATH, TOKEN, SUCCESS
} from '../../settings';

class PasswordReset extends Component {
  state = {
    values: {
      email: '',
      password: '',
      id: '',
      token: ''
    }
  };

  componentDidMount() {
    this.getResetToken();
  }

  getResetToken() {
    const query = queryString.parse(window.location.search);
    const token = Object.values(query)[0];
    if (!token) {
      setQuery({ token: '' });
      return localStorage.removeItem('token');
    }
    const decodedToken = tokenDecoder(token);
    if (decodedToken) {
      const { values } = this.state;
      const { id, email } = decodedToken;
      values.email = email;
      values.id = id;
      values.token = `Bearer ${token}`;

      return this.setState({ values });
    }
    setQuery({ token: '' });
  }

  handlePasswordChange = (event) => {
    const { name, value } = event.target;
    const { values } = this.state;
    values[name] = value;

    this.setState({ values });
  }

  resetUserPassword = async () => {
    const { resetPasswordQuery, history, dispatch } = this.props;
    const { values, values: { email } } = this.state;

    if (!email.trim()) {
      return toaster(TOASTR_ERROR, 'Check your email for password reset link');
    }

    try {
      const response = await resetPasswordQuery({
        variables: {
          ...values
        }
      });

      const { resetPassword: { message, token } } = response.data;

      toaster(SUCCESS, message);

      localStorage.setItem(TOKEN, token);
      const decodedToken = tokenDecoder(token);

      dispatch(setCurrentUser(decodedToken));
      setQuery({ token: '' });

      if (decodedToken.isVerified) {
        return history.push(MY_BOOKS_PATH);
      }

      history.push('/');
    } catch (error) {
      console.error(error);
    }
  }

  render() {
    const { values: { email } } = this.state;
    return (
      <Fragment>
        <form className="password-reset">
          <div className="form-group">
            <label
              htmlFor="exampleInputPassword1"
              className="bmd-label-floating"
            >
              Email
            </label>
            <input
              type="email"
              defaultValue={email}
              className="form-control"
              id="user-email"
              disabled
            />
          </div>
          <div className="form-group">
            <label
              htmlFor="exampleInputPassword1"
              className="bmd-label-floating"
            >
              New Password
            </label>
            <input
              type="password"
              name="password"
              className="form-control"
              id="password-reset"
              autoComplete="new-password"
              onChange={this.handlePasswordChange}
            />
            <div
              onClick={passwordToggler('password-reset')}
              id="reset-password-icon"
              role="button"
              tabIndex={0}
            >
              <ion-icon
                class="hide"
                name="eye-off"
                id="password-reset-remove-hide"
              />
              <ion-icon
                name="eye"
                id="password-reset-add-hide"
              />
            </div>
          </div>
          <div className="form-group password-reset__button">
            <button
              type="button"
              className="btn btn-primary btn-raised text-case"
              onClick={this.resetUserPassword}
            >
              Save password
            </button>
          </div>
        </form>
      </Fragment>
    );
  }
}

PasswordReset.propTypes = {
  resetPasswordQuery: PropTypes.func,
  dispatch: PropTypes.func,
  history: PropTypes.object,
};

PasswordReset.defaultProps = {
  resetPasswordQuery: () => { },
  dispatch: () => { },
  history: {},
};

export default compose(
  graphql(resetPassword, { name: RESET_PASSWORD_QUERY }),
  connect(),
)(PasswordReset);
