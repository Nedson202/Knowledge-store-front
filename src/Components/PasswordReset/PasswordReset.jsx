import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import './_PasswordReset.scss';
import passwordToggler from '../../utils/passwordToggler';
// import tokenDecoder from '../../utils/tokenDecoder';

class PasswordReset extends Component {
  componentDidMount() {
    const { history } = this.props;
    const token = window.location.search.replace(/^[?]/, '');
    history.push('password-reset');
    toastr.success('Heyey hdfh');
    return !localStorage.resetToken && localStorage.setItem('resetToken', token);
  }

  render() {
    return (
      <Fragment>
        <form className="password-reset">
          <div className="form-group">
            <label htmlFor="exampleInputPassword1" className="bmd-label-floating">
              New Password
            </label>
            <input type="password-reset" className="form-control" id="password-reset" />
            <li onClick={passwordToggler('password-reset')}>
              <i className="fa fa-eye" aria-hidden="true" id="add-hide" />
              <i className="fa fa-eye-slash hide" aria-hidden="true" id="remove-hide" />
            </li>
          </div>
          <div className="form-group password-reset__button">
            <button type="button" className="btn btn-primary btn-raised text-case">Save password</button>
          </div>
        </form>
      </Fragment>
    );
  }
}

PasswordReset.propTypes = {
  history: PropTypes.object
};

PasswordReset.defaultProps = {
  history: {}
};

export default PasswordReset;
