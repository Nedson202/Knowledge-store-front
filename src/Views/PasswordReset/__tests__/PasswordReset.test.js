import React from 'react';

import {
  render, fireEvent, AllProviders, waitForTime, cleanup
} from 'test-utils';
import { MY_BOOKS_PATH } from 'settings';

import PasswordReset from '..';
import {
  EMAIL_LABEL, NEW_PASSWORD_LABEL, TEXT, PASSWORD, PASSWORD_TOGGLE,
  SAVE_BUTTON, FORM_GROUP_CASE, PASSWORD_REST_MOCK
} from './constants';

afterEach(cleanup);

describe('PasswordReset', () => {
  it('should render form labels', () => {
    const {
      container, asFragment, getByText,
    } = render(
      <AllProviders>
        <PasswordReset />
      </AllProviders>
    );

    getByText(EMAIL_LABEL);
    getByText(NEW_PASSWORD_LABEL);

    expect(asFragment()).toMatchSnapshot();
    expect(container).toBeVisible();
  });

  it('simulate password type toggle', () => {
    const { getByTestId, getByLabelText } = render(
      <AllProviders>
        <PasswordReset />
      </AllProviders>
    );

    const passwordInput = getByLabelText(NEW_PASSWORD_LABEL);
    const passwordToggleButton = getByTestId(PASSWORD_TOGGLE);

    fireEvent.click(passwordToggleButton);
    expect(passwordInput.type).toBe(TEXT);
    fireEvent.click(passwordToggleButton);
    expect(passwordInput.type).toBe(PASSWORD);
  });

  it('should render action button', () => {
    const { getByText } = render(
      <AllProviders>
        <PasswordReset />
      </AllProviders>
    );

    getByText(SAVE_BUTTON);
  });

  it('should trigger password reset', async () => {
    window.history.pushState({}, 'Password reset', `?token=${token}`);

    const { getByText, getByLabelText } = render(
      <AllProviders apolloMocks={PASSWORD_REST_MOCK}>
        <PasswordReset />
      </AllProviders>
    );

    const savePasswordButton = getByText(SAVE_BUTTON);

    FORM_GROUP_CASE.forEach(({ name, value }) => {
      fireEvent.change(getByLabelText(name), { target: { value } });
    });

    savePasswordButton.click();

    await waitForTime(100);

    expect(localStorage.token).toBe(token);
    expect(window.location.pathname).toBe(MY_BOOKS_PATH);
  });
});
