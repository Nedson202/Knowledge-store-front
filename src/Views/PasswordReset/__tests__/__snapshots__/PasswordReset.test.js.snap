// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordReset should render form labels 1`] = `
<DocumentFragment>
  <div
    class="container-content"
  >
    <form
      class="password-reset"
    >
      <div
        class="form-group"
      >
        <label
          class="bmd-label-floating"
          for="user-email"
        >
          Email
        </label>
        <input
          class="form-control"
          disabled=""
          id="user-email"
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="bmd-label-floating"
          for="password-reset"
        >
          New Password
        </label>
        <input
          autocomplete="new-password"
          class="form-control"
          id="password-reset"
          name="password"
          type="password"
        />
        <div
          data-testid="reset-password-icon"
          id="reset-password-icon"
          role="button"
          tabindex="0"
        >
          <ion-icon
            class="hide"
            id="password-reset-remove-hide"
            name="eye-off"
          />
          <ion-icon
            id="password-reset-add-hide"
            name="eye"
          />
        </div>
      </div>
      <div
        class="form-group password-reset__button"
      >
        <button
          class="btn btn-primary btn-raised text-case"
          type="button"
        >
          Save password
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
